### PR TITLE
Initial implementation of the Shading service.

### PR DESCRIPTION
**Pull request by mcyran**

### CODE REVIEW

**Comment by rmckeown** [comment]
General question why not use the actor approach to separate shading and search as 2 separate actors. This way we can contain failures and not couple the services.

File: src/main/scala/com/bublup/search/adapters/DataAccessAdapterImpl.scala [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    * @param userId the user key
+    * @return a future entity revision user
+    */
+  override def fetchUser
+  (
+    userId: String
+  )(implicit customErrorInput: CustomErrorInput): Future[Option[EntityRevision[User]]] =
+    userGetter.get(userId)
+
+  /**
```
Comment (rmckeown) [inline src/main/scala/com/bublup/search/adapters/DataAccessAdapterImpl.scala]: nice comments explaining the chaining

File: src/main/scala/com/bublup/search/adapters/DataAccessAdapterImpl.scala [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    *                        --- (3) -->          |
+    *                                  |--- (7) -->
+    * --------------- (4) ------------->
+    *
+    * @param userId user key
+    * @param query query term
+    * @param customErrorInput custom error input
+    * @return a future search result
+    */
+  override def searchHierarchy
```
Comment (rmckeown) [inline src/main/scala/com/bublup/search/adapters/DataAccessAdapterImpl.scala]: i personally feel this should not be in the data access adapter, to me the data access adapter should be just wrappers on dal calls and not the combination of multiple dal calls with business logic

**Comment by rmckeown** [comment]
You should also sync with Ben or Matt and see what should happen with Custom error on authentication failure, I believe a custom error needs to be returned

**Comment by rmckeown** [comment]
I also do not see anything to add the custom error or extend the errors from the dal for example if the dal call fails doing shading vs search their should be different custom errors created.

**Comment by mcyran** [comment]
I'll try to answer your questions from top down:

1) I don't see the advantage of using actors over simple futures here. There is no practical difference here between handing a future failing vs an actor supervisor strategy handling an exception. Turning these into actors would be more code, more overhead, and no benefit.
2) The adapter is just a layer that converts DAL calls into objects directly consumable by the routing layer. I could split these methods into separate classes but I don't see the point.
3) CustomError is partially gutted right now because of the issue that I explained to you on Friday where it hides errors under certain circumstances. Until that gets resolved (it may have already, but I haven't see it), I'd rather see the actual exceptions thrown. Also, keep in mind that all of the dal endpoints implicitly require custom error objects, so just because you don't see it, doesn't mean custom error isn't there.

**Comment by mcyran** [comment]
I'm closing this PR and recreating in the search repo where it belongs.

**Pull request closed without merging**